web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx

```diff
@@ -0,0 +1,248 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { Dispatch, SetStateAction } from 'react';
+import { ButtonBorder, Box, Flex, Text, ButtonIcon } from 'design';
+import { Add, Trash } from 'design/Icon';
+import { FieldSelectCreatable } from 'shared/components/FieldSelect';
+import { Option } from 'shared/components/Select';
+import { requiredField, requiredAll } from 'shared/components/Validation/rules';
+import { Attempt } from 'shared/hooks/useAttemptNext';
+
+/**
+ * traitsPreset is a list of system defined traits in Teleport.
+ * The list is used to populate traits key option.
+ */
+const traitsPreset = [
+  'aws_role_arns',
+  'azure_identities',
+  'db_names',
+  'db_roles',
+  'db_users',
+  'gcp_service_accounts',
+  'host_user_gid',
+  'host_user_uid',
+  'kubernetes_groups',
+  'kubernetes_users',
+  'logins',
+  'windows_logins',
+];
+
+/**
+ * TraitsEditor supports add, edit or remove traits functionality.
+ * @param attempt attempt is Attempt status.
+ * @param configuredTraits holds traits configured for user in current editor.
+ * @param setConfiguredTraits sets user traits in current editor.
+ */
+export function TraitsEditor({
+  attempt,
+  configuredTraits,
+  setConfiguredTraits,
+}: TraitEditorProps) {
+  function handleInputChange(i: InputOption | InputOptionArray) {
+    const newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
```

Comment:
we should not allow adding empty space b/c i can do this (and save it)

we can display it if it came like that, but the web UI can prevent adding further ones

<img width="435" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/c6206f99-32a7-4eeb-8805-7b2103510157">

Reply:
I couldn't guess why would user save an empty trait value :D but I can surely avoid handing them footgun - https://github.com/gravitational/teleport/pull/42620/commits/c3b97772077a0dd6aedf919cd2e77ff1a33e5508#diff-72c47d20adf27d7885795d45757b748a86f288c7e018d043f6d0c9cac1f798ca